File: interview_ai_final/backend/app/services/gemini_service.py
import google.generativeai as genai
import json
import re
from app.core.config import settings

class GeminiService:
    def __init__(self):
        if settings.GEMINI_API_KEY:
            genai.configure(api_key=settings.GEMINI_API_KEY)
            self.model = genai.GenerativeModel('gemini-pro')
        else:
            self.model = None

    async def evaluate_answer(self, category: str, question: str, answer: str):
        if not self.model:
            return {
                "score": 0,
                "strengths": ["API Key missing"],
                "weaknesses": ["Evaluation skipped"],
                "model_answer": "Please configure Gemini API Key."
            }

        prompt = f"""
        Act as a Senior Recruiter and Interviewer for {category}.
        Evaluate the candidate's answer to the following question.

        Question: {question}
        Candidate's Answer: {answer}

        Provide feedback in JSON format exactly as follows:
        {{
            "score": <integer 0-100>,
            "strengths": [<string>, <string>, <string>],
            "weaknesses": [<string>, <string>],
            "model_answer": <string version using STAR method>
        }}

        Return ONLY the JSON object.
        """

        try:
            response = self.model.generate_content(prompt)
            text = response.text
            # Use regex to find the JSON part in case of extra text
            match = re.search(r'\{.*\}', text, re.DOTALL)
            if match:
                return json.loads(match.group())
            return json.loads(text)
        except Exception as e:
            print(f"Gemini Error: {e}")
            return {
                "score": 0,
                "strengths": ["AI Service error"],
                "weaknesses": ["Evaluation failed"],
                "model_answer": f"Error: {str(e)}"
            }

gemini_service = GeminiService()

File: interview_ai_final/lib/core/services/gemini_service.dart
import 'dart:convert';
import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:google_generative_ai/google_generative_ai.dart';

class GeminiService {
  late final GenerativeModel _model;

  GeminiService() {
    final apiKey = dotenv.env['GEMINI_API_KEY'] ?? '';
    _model = GenerativeModel(
      model: 'gemini-2.5-flash-lite',
      apiKey: apiKey,
    );
  }

  Future<String> generateInterviewQuestion({
    required String category,
  }) async {
    try {
      final response = await _model.generateContent([
        Content.text(
          'You are an expert interviewer. Generate exactly one challenging interview question for the category: "$category". '
          'Return only the plain text of the question, without any introductory or concluding remarks, explanations, or numberings.',
        )
      ]);
      return response.text?.trim() ?? 'No question generated';
    } catch (e) {
      return 'Error generating question: $e';
    }
  }

  Future<Map<String, dynamic>> evaluateAnswer({
    required String question,
    required String answer,
  }) async {
    try {
      final response = await _model.generateContent(
        [
          Content.text(
            'You are an expert interviewer. Evaluate the candidate\'s response to the following interview question.\n\n'
            'Question: "$question"\n'
            'Candidate\'s Answer: "$answer"\n\n'
            'Provide feedback strictly in JSON format with the following keys:\n'
            '- "score": an integer between 0 and 100\n'
            '- "strengths": a list of strings (2-3 items)\n'
            '- "weaknesses": a list of strings (2-3 items of areas for improvement)\n'
            '- "model_answer": a concise (3-4 sentences) exemplary answer that demonstrates how to answer this question perfectly.\n\n'
            'Your output must contain only valid JSON. Do not write any markdown code blocks or explanations outside of the JSON structure.'
          )
        ],
        generationConfig: GenerationConfig(responseMimeType: 'application/json'),
      );

      final text = response.text?.trim() ?? '{}';
      return jsonDecode(text) as Map<String, dynamic>;
    } catch (e) {
      return {
        'score': 0,
        'strengths': ['Could not evaluate response due to an error'],
        'weaknesses': ['Error details: $e'],
        'model_answer': 'N/A'
      };
    }
  }
}
